server/chat-plugins/blackjack.js

```diff
@@ -0,0 +1,643 @@
+		if (this.playerTable[user]) return this.errorMessage(user, `You don't need to spectate the game; you're playing the game.`);
+		this.spectators[user.userid] = user.userid;
+		user.sendTo(this.id, `You are now spectating this game.`);
+	}
+	unspectate(user) {
+		if (!this.spectators[user.userid]) return this.errorMessage(user, `You are already not spectating this game.`);
+		delete this.spectators[user.userid];
+		user.sendTo(this.id, `You are no longer spectating this game.`);
+	}
+
+	/**
+	 * Utility
+	 * errorMessage - sends a user an error message
+	 * add - adds/sends text to room
+	 * display - displays gameplay to players and spectators
+	 * clear - clears a user's gameplay screen
+	 * clearAllTimers - clears all possible existing timers pertaining to blackjack
+	 * slide - slides the game log down in the chat
+	 * onConnect - handles replies to send when a user joins the room, if any
+	 * onUpdateConnection - overrides default onUpdateConnection
+	 * createTimer - creates a timer with a countdown for a player
+	 * generateCard - generates the card for the UI
+	 */
+	errorMessage(user, message) {
+		user.sendTo(this.room, Chat.html`|html|<div class="message-error">${message}</div>`);
+	}
+	send(message, clean) {
+		const change = this.uhtmlChange;
+		this.room.send(`|uhtml${change}|blackjack-${this.blackjackNumber}|<div class="infobox${this.infoboxLimited}">${(clean ? message : this.lastMessage + message)}</div>`);
+		this.lastMessage += message;
+		this.uhtmlChange = 'change';
+	}
+	display(text, clean, playerName, noChange, end) {
+		let change = this.uhtmlChange;
+		if (noChange) change = '';
+		if (clean) this.lastMessage = '';
+		const message = `|uhtml${change}|blackjack-${this.blackjackNumber}|<div class="infobox${this.infoboxLimited}">`;
+		this.lastMessage += text;
+		if (end) {
+			text = `The game of blackjack has ended. <details><summary>View turn log</summary>${this.turnLog}</details>${text}`;
+			this.lastMessage = '';
+		}
+		for (let player of Object.keys(this.playerTable)) {
+			if (playerName && this.playerTable[player].name === playerName) { // turn highlighting
+				this.playerTable[player].gameLog += `<span class="highlighted">${text}</span>`;
+				this.playerTable[player].sendRoom(`${message}${end ? text : this.playerTable[player].gameLog}</div>`);
+			} else {
+				this.playerTable[player].gameLog += text;
+				this.playerTable[player].sendRoom(`${message}${end ? text : this.playerTable[player].gameLog}</div>`);
+			}
+		}
+		for (let spectator of Object.keys(this.spectators)) {
+			spectator = Users(this.spectators[spectator]);
+			if (spectator) spectator.sendTo(this.id, `${message}${this.lastMessage + text}</div>`);
+		}
+	}
+	clear() {
+		const player = this.playerTable[this.curUser];
+		if (!player) return; // this should never happen
+		player.sendRoom(`|uhtmlchange|user-blackjack-${this.blackjackNumber}|`);
+	}
+	clearAllTimers() {
+		if (this.autostand) {
```

Comment:
this doesn't seem to exist outside of this function?

Reply:
Removed